btax/front_end_util.py

```diff
@@ -0,0 +1,260 @@
+                        # all this reshaping, end up
+                        # being in the right range
+                        key1 = row['major_grouping']
+                        key2 = row['major_grouping'] == row['label']
+                        if col_key not in stats[key1][key2]:
+                            minn, maxx = stats[key1][key2][col_key] = [None, None]
+                        if minn is None or minn > value:
+                            minn = value
+                        if maxx is None or maxx < value:
+                            maxx = value
+                        stats[key1][key2][col_key] = [minn, maxx]
+                    cell = {
+                        'format': {
+                            'divisor': table['cols'][col_key]['divisor'],
+                            'decimals': table['cols'][col_key]['decimals'],
+                        },
+                        'value': value,
+                    }
+                    cell['value'] = value
+                    row['cells'].append(cell)
+                table['rows'].append(row)
+                if extra_row:
+                    # If we made a breakline row,
+                    # add it next because we are
+                    # currently processing the preceding
+                    # row
+                    table['rows'].append(extra_row)
+
+            tables[upper_key][table_id] = table
+    if do_assertions:
+        assertions_on_stats(stats)
+    tables['result_years'] = [2015]
+    return tables
+
+def _dataframe_to_json_table(df, defaults, label, index_col):
+    groups = [x[1]['table_id'] for x in defaults]
+    tables = defaultdict(lambda: {})
+    for group in set(groups):
+        if group == 'all':
+            continue
+        new_column_names = [x[1]['col_label'] for x in defaults
+                            if x[1]['table_id'] == group]
+        keep_columns = [x[0] for x in defaults
+                        if x[1]['table_id'] in (group, 'all')]
+        df2 = df[keep_columns]
+        df2[index_col] = [replace_unicode_spaces(s) for s in df2[index_col]]
+        df2.set_index(index_col, inplace=True)
+        df2.columns = new_column_names
+        if 'reform' in label:
+            label2 = 'reform'
+        elif 'changed' in label:
+            label2 = 'changed'
+        elif 'base' in label:
+            label2 = 'baseline'
+        print(label, label2, group)
+        tables[group][label2] = df2
+    return tables
+
+def output_by_asset_to_json_table(df, table_name):
+    from btax.parameters import DEFAULT_ASSET_COLS
+    return _dataframe_to_json_table(df, DEFAULT_ASSET_COLS,
+                                    table_name, 'Asset Type')
+
+
+def output_by_industry_to_json_table(df, table_name):
+    from btax.parameters import DEFAULT_INDUSTRY_COLS
+    return _dataframe_to_json_table(df, DEFAULT_INDUSTRY_COLS,
+                                    table_name, 'Industry')
+
+
+
+def assertions_on_stats(stats):
```

Comment:
@talumbau @jdebacker This is a new set of assertions to confirm that the bold summary rows are within the min / max range of the rows they are supposed to summarize (the summaries should be weighted means).  If `do_assertions=True`, then `stats` are accumulated in `add_summary_rows_and_breaklines` for checking with `assertions_on_stats`.  `do_assertions` is controlled by environment variable `BTAX_TABLE_ASSERTIONS` (1 or 0) so we can even run the assertions in live testing.  The test module in btax/tests runs btax a number of times passing `do_assertions=True` indirectly.  I have confirmed the tests are fixed (recall not all are run in Travis):

```
$ BTAX_CUR_DIR=`pwd`/btax py.test btax/tests
============================= test session starts ==============================
platform darwin -- Python 2.7.10, pytest-3.0.3, py-1.4.31, pluggy-0.4.0
rootdir: /Users/psteinberg/Documents/AEI/btax_new, inifile: setup.cfg
plugins: django-3.0.0, pep8-1.0.6
collected 39 items

btax/tests/test_params_have_effect.py .......................................

============================ pytest-warning summary ============================
WC1 None [pytest] section in setup.cfg files is deprecated, use [tool:pytest] instead.
=============== 39 passed, 1 pytest-warnings in 1007.33 seconds ========
```

Let me know if there are more assertions to do.  I'll make another line comment where other assertions on table format are done.